ENH: Allow, and default to, downstream building with old API

seberg opened the pull request:
This implements gh-23525.  I made the slightly more conservative choice of not messing with our C version numbering after realizing that we have the `NPY_1_23_API_VERSION` macros anyway.

So this adds the ability to do:
```
#define NPY_TARGET_VERSION NPY_1_22_API_VERSION
```
that defaults to `NPY_1_17_API_VERSION` (since that is reasonable) and I allow to go back up to 1.15.

I did not bother to hide all structs (if they are unusable because all accessor structs are hidden).  The most relevant change that I found looking through commit history and release notes that isn't captured is this one:
https://github.com/numpy/numpy/pull/14393/files

Which is arguably always better, but of course compatibility is in theory not perfect (in both directions really!).

TODOs:
* [ ] feedback?
* [x] release notes
* [x] Brief user documentation (in the document linked also)
* [x] Brief point in the contributor guide since adding guards and `MinVersion` is very important now.

I have compiled SciPy and run on an older version, although not old enough to be relevant yet (M1 just doesn't support much back).  OTOH, it clearly tests that the target mechanism works.

seberg committed:
BLD: Add support for NPY_TARGET_VERSION macro

This is a way for downstream users to specify which NumPy version
they wish to be compaible with.

Note that we provide a conservative default here (because almost nobody
actually uses new API as they would lose backwards compatibility).

Initially I had thought we should just redefine it so that the target
version uses the same scheme as the Python hex version (and limited API),
but since we have `NPY_1_15_API_VERSION` defines, use those...

This commit does not include any actual use of this!

seberg committed:
ENH: Allow compiling compatibly to old NumPy versions

The default compiles compatibly with 1.17.x, we allow going back to
1.15 (mainly because it is easy).

There were few additions in this time, a few structs grew and very
few API functions were added.  Added a way to mark API functions
as requiring a specific target version.

If a user wishes to use the *new* API, they have to add the definition:

    #define NPY_TARGET_VERSION NPY_1_22_API_VERSION

Before importing NumPy.  (Our version numbering is a bit funny
I first thought to use a hex version of the main NumPy version,
but since we already have the `NPY_1_22_API_VERSION` defines...)

seberg committed:
MAINT: Bump version in meson

seberg committed:
BUG: Fix typo in new version checks

seberg commented on numpy/core/code_generators/cversions.txt:
```diff
@@ -67,5 +67,7 @@
 0x00000010 = 04a7bf1e65350926a0e528798da263c0
+
+# Version 17 (NumPy 1.25) No actual change.
+0x00000011 = ca1aebdad799358149567d9d93cbca09
```
The min-version macros do goes into the hash right now.  So removing them will also change it, but that seems fine?

rgommers replied:
Only for >= 1.25, right? Should be fine I'd think.

seberg committed:
TST,DOC: Avoid spaces to hopefully ensure more info on error and fix memtests

seberg commented on numpy/core/code_generators/genapi.py:
```diff
@@ -98,6 +99,33 @@ def _repl(str):
+        """Wrap a definition behind a version guard"""
+        numpy_target_help_pointer = (
+            "(Old_NumPy_target_see_depending_on_numpy__"
+            "https://numpy.org/devdocs/dev/depending_on_numpy.html)")
```
Maybe its better to just not define it and let people google.  It looked pretty decent when I tried it in godbolt, but when the memory handler tests failed now, it didn't work well (no spaces might help, but I suspect not, since the `:` and `/` will mess things up.

rgommers replied:
seems reasonable to leave this out

mattip commented:
> I have compiled SciPy

How did you pass the define to meson?

seberg commented:
I didn't pass anything, so it would default to 1.17, which is what SciPy would want to use.

seberg commented on numpy/core/tests/test_mem_policy.py:
```diff
@@ -89,6 +89,7 @@ def get_module(tmp_path):
     prologue = '''
+        #define NPY_TARGET_VERSION NPY_1_22_API_VERSION
         #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
         #include <numpy/arrayobject.h>
```
@mattip this is also a nice confirmation.  Without this define build fails for the memory policy tests, since it requires API newer than 1.17.

mattip commented:
Ahh, the default in this PR to use `NPY_1_17_API_VERSION`. So going forward anyone who wants to use 1.25 locally will need to use a define, specifically `#define NPY_TARGET_VERSION NPY_1_25_API_VERSION. This is a change. Until now it was sufficient to install NumPy 1.25 (or list it as a dependency) to get 1.25 semantics. Going forward, everyone will need to define the desired API version in addition to installing the version.

seberg commented:
Yes, if you want to use the newest API you will have to define `NPY_TARGET_VERSION NPY_API_VERSION`.  I don't think this matters in practice (at this point), because there is only one relevant change in the API (the memory handler) and that is very niche.

Overall, I honestly think this is a big improvement because those who use cython and distribute wheels will have _not_ have to worry about it.

seberg committed:
Tweak to just warn for low versions

seberg committed:
DOC: Start on docs about compatible downstream builds

This still needs to be expanded and maybe removes a bit more than
it should.  OTOH, the actual changes will mainly be necessary once
NumPy 2.0 comes, right now the old scheme remains valid.

seberg committed:
DOC: Add API change section to reviewer docs

This seemed like the clearest place to add a more in-depth note
on it.  The actual table, etc. also would make sense, but there
you probably see the `MinVersion` anyway...

rgommers commented on doc/release/upcoming_changes/23528.compatibility.rst:
```diff
@@ -0,0 +1,15 @@
+Default build against NumPy is now backwards compatible
```
This section header may need a tweak to avoid reader confusion, maybe "Building against the NumPy C API now yields binaries that work with older NumPy versions"

rgommers commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -54,64 +54,67 @@ Adding a dependency on NumPy
+    have to compile with the oldest version you wish to support.
+    This can be done by using
+    `oldest-supported-numpy <https://github.com/scipy/oldest-supported-numpy/>`__.
+    Please see the NumPy 1.24 documentation for further details.
```
Link to https://numpy.org/doc/1.24/dev/depending_on_numpy.html directly?

rgommers commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -54,64 +54,67 @@ Adding a dependency on NumPy
+oldest NumPy version that supports the currently oldest compatible Python
+version:  NumPy 1.25.0 supports Python 3.9 and higher and NumPy 1.19 is the
+first version to support Python 3.9.  Thus, we will guarantee that version
+(the exact version is set in the headers).
```
minor tweak proposed: "is set within NumPy-internal header files"

rgommers commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -54,64 +54,67 @@ Adding a dependency on NumPy
+(the exact version is set in the headers).
 
-.. note::
+NumPy is also forward compatible for all minor release, but a major release
```
typo: add `s` after `all minor release`

rgommers commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -54,64 +54,67 @@ Adding a dependency on NumPy
-    ``conda`` will always use ``-no-build-isolation``; dependencies for conda
-    builds are given in the conda recipe (``meta.yaml``), the ones in
-    ``pyproject.toml`` have no effect.
+before including any NumPy headers (or the equivalent ``-D`` compiler flag).
```
I briefly wondered whether it matters whether this define is used consistently globally or per extension module. I believe it's the latter; would be useful to clarify that here.

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -101,7 +101,34 @@ For maintainers
+API Changes
+-----------
+As mentioned most public API changes should be discussed ahead of time and
+often with a wider audience (mailinglist or even through a NEP).
```
typo in "mailinglist". I propose "on the mailing list, or even through a NEP"

rgommers reviewed (commented):
This looks quite good to me, thanks @seberg!

The main questions/comments I have are:
- It may be useful to say something about the new `DType` related C API, and why it does or doesn't matter here.
- Compilers and compiler versions: I'm wondering if we are making stronger assumptions about whether the older and latest numpy versions are built with the same (or a compatible) compiler versions
  - I think at least this becomes more of a problem with the use of `libnpymath`/`libnpyrandom`, and we'll probably have to get rid of those before starting to use this in SciPy?

rgommers commented:
I haven't tested yet, but will make some time to play with this.

seberg committed:
MAINT,DOC: Update based on Ralf's review

seberg commented:
> It may be useful to say something about the new DType related C API, and why it does or doesn't matter here.

The new DType API isn't public yet and much more restrictive (no compat guarantee), so I don't see that it needs mentioning?  Once release, it would be hidden behind the versioning (mostly).

> I think at least this becomes more of a problem with the use of `libnpymath`/`libnpyrandom`

Hmmm, I figured as static libs that shouldn't use numpy proper we don't have to worry too much.
Is downstream ever using old compilers intentionally for compat with old NumPy versions?  It seems like possibly important, but I can't think of where it matters.  (Although, I would probably worry more if it was C++ API.)

rgommers commented:
> The new DType API isn't public yet

Ah I missed that - never mind then:)

> Is downstream ever using old compilers intentionally for compat with old NumPy versions? 

Yes, it happened at least once, on Windows when the vc141 and vc142 toolchains weren't compatible. It was a corner case though. In general, I'd say using the same compilers for building your whole stack is important either way though. So I can imagine cases where you build against numpy 1.25.0, which then gives you a C API that's compatible with >=1.19.0 but a static `libnpymath.so` that isn't. 

Let's not worry about that too much here though, we should just go ahead with deprecating those static libraries and stop using them.

seberg committed:
DOC: Align default with what the NEP draft says

We have three choices:
1. Be compatible with limited API (oldest supported Python version)
2. Be compatible with everything on the same Python version (this)
3. Be strict and default to NEP 29

This just rephrases things to be 2.  Because our API version was not
bumped over the relevant time frame, it is actually also compatible
with the first.

mattip commented on doc/release/upcoming_changes/23528.compatibility.rst:
```diff
@@ -0,0 +1,16 @@
+Binaries build against the NumPy C API now with older NumPy versions
+--------------------------------------------------------------------
```
Maybe
```suggestion
By default, the exported NumPy C API is now compatible with NumPy 1.17
----------------------------------------------------------------------
```

seberg replied:
This meandered a bit.  As far as I guessed, `oldest-support-numpy` would pick `1.19` for NumPy 1.25, so that seemed like the right thing now.  But 1.16-1.19 all are the same API version (since we didn't always bump it in the past).

I previously had written 1.17 everywhere though.  The question is what to write here now, 1.19 as what we would pick, or 1.16 because that is what we end up with anyway?

@rgommers 1.19 was correct for `oldest-support-numpy`?  (because 1.19.x is the first version to support Python 3.9)

rgommers replied:
It doesn't seem to matter indeed. Yes, `1.19.3` is what the default is for `oldest-supported-numpy`. I like 1.19, it just eliminated versions to worry about that are irrelevant anyway.

mattip commented on doc/release/upcoming_changes/23528.compatibility.rst:
```diff
@@ -0,0 +1,16 @@
+Starting with NumPy 1.25 when including NumPy headers, NumPy now
+defaults to exposing a backwards compatible API.
+This means that by default binaries such as wheels build against
+NumPy 1.25 will also work with NumPy 1.16 because it has the same API version
```
Not 1.17?
```suggestion
NumPy 1.25 will also work with NumPy 1.17 because it has the same API version
```

mattip commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -48,70 +48,78 @@ job, either all warnings or otherwise at least ``DeprecationWarning`` and
-      - numpy
-    run:
-      - {{ pin_compatible('numpy') }}
+By default, NumPy will expose an API that is compatible with at least the
+oldest NumPy version that supports the currently oldest compatible Python
```
```suggestion
By default, NumPy will expose an API that is backwardscompatible with the
oldest NumPy version that supports the currently oldest compatible Python
```

mattip commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -48,70 +48,78 @@ job, either all warnings or otherwise at least ``DeprecationWarning`` and
-      - {{ pin_compatible('numpy') }}
+By default, NumPy will expose an API that is compatible with at least the
+oldest NumPy version that supports the currently oldest compatible Python
+version:  NumPy 1.25.0 supports Python 3.9 and higher and NumPy 1.19 is the
```
```suggestion
version.  NumPy 1.25.0 supports Python 3.9 and higher and NumPy 1.19 is the
```

mattip commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -48,70 +48,78 @@ job, either all warnings or otherwise at least ``DeprecationWarning`` and
+By default, NumPy will expose an API that is compatible with at least the
+oldest NumPy version that supports the currently oldest compatible Python
+version:  NumPy 1.25.0 supports Python 3.9 and higher and NumPy 1.19 is the
+first version to support Python 3.9.  Thus, we will guarantee that version
```
```suggestion
first version to support Python 3.9.  Thus, we guarantee that version
```

mattip commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -48,70 +48,78 @@ job, either all warnings or otherwise at least ``DeprecationWarning`` and
-    ``conda`` will always use ``-no-build-isolation``; dependencies for conda
-    builds are given in the conda recipe (``meta.yaml``), the ones in
-    ``pyproject.toml`` have no effect.
+before including any NumPy headers (or the equivalent ``-D`` compiler flag).
```
```suggestion
before including any NumPy headers (or the equivalent ``-D`` compiler flag) in
every extension module that requires the NumPy C-API.
```

mattip commented on doc/source/dev/depending_on_numpy.rst:
```diff
@@ -48,70 +48,78 @@ job, either all warnings or otherwise at least ``DeprecationWarning`` and
-    upper version bound in your build configuration for the last known version
-    of ``setuptools`` that works with your build.
+    At the time of NumPy 1.25, NumPy 2.0 is expected to be the next release
+    of NumPy.  The NumPy 2.0 release is expected to require modifying this
+    since it will be required to compile against NumPy 2+ in order to be
+    compatible with both NumPy 1.x and 2.x.
```
```suggestion
    of NumPy.  The NumPy 2.0 release is expected to require a different pin,
    since NumPy 2+ will be needed in order to be compatible with both NumPy
    1.x and 2.x.
```

mattip commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -101,7 +101,34 @@ For maintainers
+
+    'PyDataMem_SetHandler':                 (304, MinVersion("1.22")),
+
+Header only functionality (such as a new macro) typically do not need to be
```
```suggestion
Header only functionality (such as a new macro) typically does not need to be
```

mattip commented on numpy/core/setup_common.py:
```diff
@@ -48,7 +48,13 @@
+# 0x00000011 - 1.25.x
+C_API_VERSION = 0x00000011
+
+# When compiling against NumPy (downstream libraries), NumPy will by default
```
The "by default" is the important part, make sure it comes first

```suggestion
# By default, when compiling downstream libraries against NumPy,```

mattip commented:
Some comment and docstring suggestions, I left it to you to decide if they are improvements or not. Other than that this looks ready to go.

seberg commented on numpy/core/setup_common.py:
```diff
@@ -48,7 +48,13 @@
+
+# When compiling against NumPy (downstream libraries), NumPy will by default
+# pick an older feature version.  For example, for 1.25.x we default to the
+# 1.17 API and support going back all the way to 1.15.x (if so desired).
```
```suggestion
# 1.19 API and support going back all the way to 1.15.x (if so desired).
```

seberg committed:
Apply suggestions from code review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

mattip commented:
Thanks @seberg
